content/blog/2022/2022-01-16-spring-boot-null-safety-annotations.md

```diff
@@ -0,0 +1,152 @@
+
+// imports 
+
+class Employee {
+  String id;
+  String name;
+  LocalDate joiningDate;
+  String pastEmployment;
+
+  // standard constructor, getters, setters
+}
+```
+
+Mostly the `id` field is going to be a non-nullable value. So, to avoid any potential `NullPointerException` we can mark this field as [`NonNull`](https://docs.spring.io/spring-framework/docs/current/javadoc-api/org/springframework/lang/NonNull.html):
+
+
+```java
+class Employee {
+
+  @NonNull 
+  String id;
+
+  //...
+}
+```
+Now, if we accidentally try to set the value of `id` as null anywhere in the code, the IDE will show a compilation warning:
+
+{{% image alt="IDE warning for NonNull" src="images/posts/spring-boot-null-safety-annotations/nonnull-ide-warning.png" %}}
+
+Pretty awesome, isn't it?
+
+**The `NonNull` annotation can be used at the method, parameter, or field level.** 
+
+At this point, you might be thinking, what if a class has more than one non-null field. Would it not be too wordy if we have to add a `@NonNull` annotation before each of these? 
+
+We can solve this problem by using the [`@NonNullFields`](https://docs.spring.io/spring-framework/docs/current/javadoc-api/org/springframework/lang/NonNullFields.html) annotation. Let us see how in the next section.
+
+## `@NonNullFields`
+
+Let us create a `package-info.java` file to apply the non-null field checks at the package level. This file will contain the root package name with `@NonNullFields` annotation:
+
+```java
+@NonNullFields
+package io.reflectoring.nullsafety;
+
+import org.springframework.lang.NonNullFields;
+```
+
+Now, **we no longer need to annotate the fields with `@NonNull`**. Because by default, all the fields are now treated as non-null. Thanks to `@NonNullFields`! And, we will still see the same warning as before:
+
+{{% image alt="IDE warning for NonNullFields" src="images/posts/spring-boot-null-safety-annotations/nonnull-ide-warning.png" %}}
+
+Another point to note here is if there are any uninitialized fields, then we will see a warning to initialize those:
+
+{{% image alt="IDE warning for NonNull" src="images/posts/spring-boot-null-safety-annotations/nonnullfields-ide-warning.png" %}}
+
+## `@NonNullApi`
+
+By now, you might have spotted another requirement, i.e., to have similar checks for method parameters or return values. Here [`@NonNullApi`](https://docs.spring.io/spring-framework/docs/current/javadoc-api/org/springframework/lang/NonNullApi.html) will come to our rescue. 
+
+In the same fashion, as we did for `@NonNullFields`, we can use a `package-info.java` file and add the `@NonNullApi` annotation for the intended package:
+
+```java
+@NonNullApi
+package io.reflectoring.nullsafety;
+
+import org.springframework.lang.NonNullApi;
+```
+
+We can now see the IDE is warning us about the non-nullable parameters:
+
+{{% image alt="IDE warning for NonNullApi" src="images/posts/spring-boot-null-safety-annotations/nonnullapi-ide-warning.png" %}}
```

Comment:
Please add a code example that shows how you use the `@NonNullApi` annotation. From just the warning I can guess how it works, but I don't know for sure :)

Reply:
Fair enough. I have now updated this section.